test/integration/src/test/java/io/pravega/test/integration/KeyValueTableTest.java

```diff
@@ -22,63 +22,98 @@
     private static final String SCOPE = "Scope";
     private static final KeyValueTableConfiguration DEFAULT_CONFIG = KeyValueTableConfiguration.builder().partitionCount(5).build();
     private static final Duration TIMEOUT = Duration.ofSeconds(30);
     private ServiceBuilder serviceBuilder;
     private TableStore tableStore;
-    private PravegaConnectionListener serverListener;
+    private PravegaConnectionListener server = null;
     private ConnectionFactory connectionFactory;
+    private TestingServer zkTestServer = null;
+    private ControllerWrapper controllerWrapper = null;
     private Controller controller;
     private KeyValueTableFactory keyValueTableFactory;
 
+
     @Before
     public void setup() throws Exception {
         super.setup();
+
+        final int controllerPort = TestUtils.getAvailableListenPort();
+        final String serviceHost = ENDPOINT;
+        final int servicePort = TestUtils.getAvailableListenPort();
+        final int containerCount = 4;
+
+        // 1. Start ZK
+        this.zkTestServer = new TestingServerStarter().start();
+
+        // 2. Start Pravega SegmentStore service.
         this.serviceBuilder = ServiceBuilder.newInMemoryBuilder(ServiceBuilderConfig.getDefaultConfig());
-        this.serviceBuilder.initialize();
-        this.tableStore = this.serviceBuilder.createTableStoreService();
-        int port = TestUtils.getAvailableListenPort();
-        this.serverListener = new PravegaConnectionListener(false, port, mock(StreamSegmentStore.class), this.tableStore);
-        this.serverListener.startListening();
+        serviceBuilder.initialize();
+        StreamSegmentStore store = serviceBuilder.createStreamSegmentService();
+        this.tableStore = serviceBuilder.createTableStoreService();
 
-        this.connectionFactory = new ConnectionFactoryImpl(ClientConfig.builder().build());
-        this.controller = new MockController(ENDPOINT, port, this.connectionFactory, true);
+        this.server = new PravegaConnectionListener(false, servicePort, store, tableStore);
+        this.server.startListening();
+
+        // 3. Start Pravega Controller service
+        this.controllerWrapper = new ControllerWrapper(zkTestServer.getConnectString(), false,
+                controllerPort, serviceHost, servicePort, containerCount);
+        this.controllerWrapper.awaitRunning();
+        this.controller = controllerWrapper.getController();
+
+        //4. Create Scope
         this.controller.createScope(SCOPE);
+        this.connectionFactory = new ConnectionFactoryImpl(ClientConfig.builder().build());
         this.keyValueTableFactory = new KeyValueTableFactoryImpl(SCOPE, this.controller, this.connectionFactory);
     }
 
+
     @After
-    public void tearDown() {
+    public void tearDown() throws Exception {
         this.controller.close();
         this.connectionFactory.close();
-        this.serverListener.close();
+
+        if (this.controllerWrapper != null) {
```

Comment:
Why did you use this pattern here and not apply it to everything else? For cleanliness, I suggest you don't bother with this check.

Reply:
ok removed.